vulcanai/models/basenetwork.py

```diff
@@ -872,32 +873,36 @@ def forward_pass(self, data_loader, transform_outputs=False,
         # prediction_shape used to aggregate network outputs
         # (e.g. with or without class conversion)
         # so far always a float.
-        dtype = torch.float
-        pred_collector = torch.tensor([], dtype=dtype, device=self.device)
+        pred_collector = None
         for data, _ in data_loader:
+            data = set_tensor_device(data, device=self.device)
             # Get raw network output
-            raw_outputs = self(data)
+            raw_outputs = self(data).cpu().detach().numpy()
             if self._num_classes:
                 if self._final_transform:
                     predictions = self._final_transform(raw_outputs)
```

Comment:
Wouldn't it be better to keep the torch tensor through the final transform?

Reply:
because it could be a numpy function, no?

Reply:
I guess it could be but this is where we’d apply a soft max for instance and maybe it’s better to keep as torch for optimization?

Reply:
Only in the case of nn.CrossEntropyLoss?

Reply:
perhaps not only for CrossEntropy but should we restrict users to only Pytorch functions or numpy functions? curious to hear your thoughts on this - I think it really depends on the scope of those final transforms and how complex but still

Reply:
uhhhhhhhmmmmmm yeah I don't really know. Personally, I think we should choose ease of use over speed, particularly as in this case I don't see a huge GPU advantage? it's a transform after training.

Reply:
like, with it this way they can give either kind of function. I definitely want to be able to pass something like numpy.round() or whatever.

Reply:
You're right - it won't slow down training but then depending on the function passed it may slow it down during inference. To put to ease the speed issue, would you be able to experiment the numpy.round() example and compared with a comparative torch.round() and see what, if any, slow down there is (and potentially for softmax)?